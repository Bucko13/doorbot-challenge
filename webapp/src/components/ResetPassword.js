import React from 'react';
import PropTypes from 'prop-types';
import { Message, Button, Form, Header, Segment } from 'semantic-ui-react'

const ResetPassword = ({ form, onChangePassword, onSubmit, error }) => {
  return (
    <Segment size='tiny' padded='very'>
      <Header as='h1'>Please set a new password</Header>
      { error ?
        <Message
          data-test='error'
          error
          header='Failed to reset password'
          content={error}
        /> :
          null
      }
      <Form onSubmit={onSubmit} data-test='form'>
        <Form.Field>
          <label>New Password</label>
          <input
            type='password'
            onChange={onChangePassword}
            data-test='password'
            value={form.password}
            placeholder='Password'
          />
        </Form.Field>
        <Button color='teal' type='submit'>Set new password</Button>
      </Form>
    </Segment>
  )
}

ResetPassword.propTypes = {
  form: PropTypes.shape({
    password: PropTypes.string
  }),
  onChangePassword: PropTypes.func.isRequired,
  onSubmit: PropTypes.func.isRequired,
  error: PropTypes.string,
};

export default ResetPassword;
