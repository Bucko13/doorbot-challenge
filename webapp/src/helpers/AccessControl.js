import React from 'react';
import { Redirect } from 'react-router-dom';
import { connect } from 'react-redux';

function AccessControl({ userRole, user, component: Component }) {
  if (userRole === 'user' && user.username) {
    return <Component />;
  }

  if (userRole === 'admin' && user.admin) {
    return <Component />;
  }

  return <Redirect to='/login' />
}

const mapStateToProps = function(state) {
  return {
    user: state.user
  }
}

const ConnectedAccessControl = connect(mapStateToProps)(AccessControl);

export function ifLoggedIn(Component) {
  return (
    () => <ConnectedAccessControl userRole={'user'} component={Component}/>
  );
}

export function ifAdmin(Component) {
  return (
    () => <ConnectedAccessControl userRole={'admin'} component={Component}/>
  )
}
