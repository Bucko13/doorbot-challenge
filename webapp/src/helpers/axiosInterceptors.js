export function straigthResponseData(response) {
  return response.data;
}

export function straigthResponseError(error) {
  return Promise.reject(error.response.data);
}
