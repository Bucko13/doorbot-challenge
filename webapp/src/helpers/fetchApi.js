// Small library to improve on fetch() usage
const fetchApi = function(method, url, data, headers = {}){
  return fetch(url, {
    method: method.toUpperCase(),
    body: JSON.stringify(data),  // send it as stringified json
    credentials: fetchApi.credentials,  // to keep the session on the request
    headers: Object.assign({}, fetchApi.headers, headers)  // extend the headers
  }).then(res => {
    if (res.ok) { return res.json() }

    const contentType = res.headers.get("content-type");
    if (contentType && contentType.indexOf("application/json") !== -1) {
      return res.json().then((data) => Promise.reject(data));
    } else {
      return Promise.reject({ error: 'There was an internal server error. Please try again.' });
    }
  });
};

// Defaults that can be globally overwritten
fetchApi.credentials = 'include';
fetchApi.headers = {
  'csrf-token': window.csrf || '',    // only if globally set, otherwise ignored
  'Accept': 'application/json',       // receive json
  'Content-Type': 'application/json'  // send json
};

// Convenient methods
['get', 'post', 'put', 'delete'].forEach(method => {
  fetchApi[method] = fetchApi.bind(null, method);
});

export default fetchApi;
