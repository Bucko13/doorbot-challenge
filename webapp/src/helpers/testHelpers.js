export function sel(id) {
  return `[data-test="${id}"]`;
}

export function changeInputValue(input, value) {
  input.simulate('change', { target: value });
}
