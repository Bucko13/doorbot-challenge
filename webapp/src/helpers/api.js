import fetchApi from './fetchApi';

function login({ username, password }) {
  return fetchApi.post('/api/v1/auth', { username, password });
}

const api = {
  login
}

export default api;
