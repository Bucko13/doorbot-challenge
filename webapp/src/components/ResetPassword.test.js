import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import ResetPassword from './ResetPassword';
import { sel, changeInputValue } from '../helpers/testHelpers';

describe('ResetPassword Component', () => {
  it('starts with an empty form', () => {
    const wrapper = generateComponent();
    expect(wrapper).toMatchSnapshot();
  });

  it('fills in password', () => {
    const onChangePassword = jest.fn();
    const wrapper = generateComponent({ onChangePassword });
    const input = wrapper.find(sel('password'));

    changeInputValue(input, 'admin');

    expect(onChangePassword.mock.calls.length).toBe(1);
  });

  it('submits resets password form', () => {
    const onSubmit = jest.fn();
    const wrapper = generateComponent({ onSubmit });
    const form = wrapper.find(sel('form'));

    form.simulate('submit');

    expect(onSubmit.mock.calls.length).toBe(1);
  });

  it('displays error', () => {
    const error = 'Invalid username or password';
    const wrapper = generateComponent({ error });
    const errorMessage = wrapper.find(sel('error'));

    expect(errorMessage.render().text()).toContain(error);
    expect(errorMessage).toMatchSnapshot();
  });
});

function generateComponent(props = {}) {
  const propsToUse = {
    form: {
      password: '',
    },
    onChangePassword: () => {},
    onSubmit: () => {},
    ...props,
  }

  return shallow(<ResetPassword {...propsToUse} />);
}
