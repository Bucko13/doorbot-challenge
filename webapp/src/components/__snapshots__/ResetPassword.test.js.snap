// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResetPassword Component displays error 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): ShallowWrapper {
    "length": 1,
    Symbol(enzyme.__root__): [Circular],
    Symbol(enzyme.__unrendered__): <ResetPassword
      error="Invalid username or password"
      form={
            Object {
                  "password": "",
                }
      }
      onChangePassword={[Function]}
      onSubmit={[Function]}
/>,
    Symbol(enzyme.__renderer__): Object {
      "batchedUpdates": [Function],
      "getNode": [Function],
      "render": [Function],
      "simulateEvent": [Function],
      "unmount": [Function],
    },
    Symbol(enzyme.__node__): Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <Header
            as="h1"
>
            Please set a new password
</Header>,
          <Message
            content="Invalid username or password"
            data-test="error"
            error={true}
            header="Failed to reset password"
/>,
          <Form
            as="form"
            data-test="form"
            onSubmit={[Function]}
>
            <FormField>
                        <label>
                                    New Password
                        </label>
                        <input
                                    data-test="password"
                                    onChange={[Function]}
                                    placeholder="Password"
                                    type="password"
                                    value=""
                        />
            </FormField>
            <Button
                        as="button"
                        color="teal"
                        type="submit"
            >
                        Set new password
            </Button>
</Form>,
        ],
        "padded": "very",
        "size": "tiny",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "as": "h1",
            "children": "Please set a new password",
          },
          "ref": null,
          "rendered": "Please set a new password",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "content": "Invalid username or password",
            "data-test": "error",
            "error": true,
            "header": "Failed to reset password",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "as": "form",
            "children": Array [
              <FormField>
                <label>
                                New Password
                </label>
                <input
                                data-test="password"
                                onChange={[Function]}
                                placeholder="Password"
                                type="password"
                                value=""
                />
</FormField>,
              <Button
                as="button"
                color="teal"
                type="submit"
>
                Set new password
</Button>,
            ],
            "data-test": "form",
            "onSubmit": [Function],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": Array [
                  <label>
                    New Password
</label>,
                  <input
                    data-test="password"
                    onChange={[Function]}
                    placeholder="Password"
                    type="password"
                    value=""
/>,
                ],
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "New Password",
                  },
                  "ref": null,
                  "rendered": "New Password",
                  "type": "label",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "data-test": "password",
                    "onChange": [Function],
                    "placeholder": "Password",
                    "type": "password",
                    "value": "",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": "input",
                },
              ],
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "as": "button",
                "children": "Set new password",
                "color": "teal",
                "type": "submit",
              },
              "ref": null,
              "rendered": "Set new password",
              "type": [Function],
            },
          ],
          "type": [Function],
        },
      ],
      "type": [Function],
    },
    Symbol(enzyme.__nodes__): Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": Array [
            <Header
              as="h1"
>
              Please set a new password
</Header>,
            <Message
              content="Invalid username or password"
              data-test="error"
              error={true}
              header="Failed to reset password"
/>,
            <Form
              as="form"
              data-test="form"
              onSubmit={[Function]}
>
              <FormField>
                            <label>
                                          New Password
                            </label>
                            <input
                                          data-test="password"
                                          onChange={[Function]}
                                          placeholder="Password"
                                          type="password"
                                          value=""
                            />
              </FormField>
              <Button
                            as="button"
                            color="teal"
                            type="submit"
              >
                            Set new password
              </Button>
</Form>,
          ],
          "padded": "very",
          "size": "tiny",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "as": "h1",
              "children": "Please set a new password",
            },
            "ref": null,
            "rendered": "Please set a new password",
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "content": "Invalid username or password",
              "data-test": "error",
              "error": true,
              "header": "Failed to reset password",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "as": "form",
              "children": Array [
                <FormField>
                  <label>
                                    New Password
                  </label>
                  <input
                                    data-test="password"
                                    onChange={[Function]}
                                    placeholder="Password"
                                    type="password"
                                    value=""
                  />
</FormField>,
                <Button
                  as="button"
                  color="teal"
                  type="submit"
>
                  Set new password
</Button>,
              ],
              "data-test": "form",
              "onSubmit": [Function],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "children": Array [
                    <label>
                      New Password
</label>,
                    <input
                      data-test="password"
                      onChange={[Function]}
                      placeholder="Password"
                      type="password"
                      value=""
/>,
                  ],
                },
                "ref": null,
                "rendered": Array [
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": "New Password",
                    },
                    "ref": null,
                    "rendered": "New Password",
                    "type": "label",
                  },
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "data-test": "password",
                      "onChange": [Function],
                      "placeholder": "Password",
                      "type": "password",
                      "value": "",
                    },
                    "ref": null,
                    "rendered": null,
                    "type": "input",
                  },
                ],
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "as": "button",
                  "children": "Set new password",
                  "color": "teal",
                  "type": "submit",
                },
                "ref": null,
                "rendered": "Set new password",
                "type": [Function],
              },
            ],
            "type": [Function],
          },
        ],
        "type": [Function],
      },
    ],
    Symbol(enzyme.__options__): Object {
      "adapter": ReactSixteenAdapter {
        "options": Object {
          "enableComponentDidUpdateOnSetState": true,
        },
      },
    },
  },
  Symbol(enzyme.__unrendered__): null,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "content": "Invalid username or password",
      "data-test": "error",
      "error": true,
      "header": "Failed to reset password",
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "content": "Invalid username or password",
        "data-test": "error",
        "error": true,
        "header": "Failed to reset password",
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`ResetPassword Component starts with an empty form 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ResetPassword
    form={
        Object {
            "password": "",
          }
    }
    onChangePassword={[Function]}
    onSubmit={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <Header
          as="h1"
>
          Please set a new password
</Header>,
        null,
        <Form
          as="form"
          data-test="form"
          onSubmit={[Function]}
>
          <FormField>
                    <label>
                              New Password
                    </label>
                    <input
                              data-test="password"
                              onChange={[Function]}
                              placeholder="Password"
                              type="password"
                              value=""
                    />
          </FormField>
          <Button
                    as="button"
                    color="teal"
                    type="submit"
          >
                    Set new password
          </Button>
</Form>,
      ],
      "padded": "very",
      "size": "tiny",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "as": "h1",
          "children": "Please set a new password",
        },
        "ref": null,
        "rendered": "Please set a new password",
        "type": [Function],
      },
      null,
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "as": "form",
          "children": Array [
            <FormField>
              <label>
                            New Password
              </label>
              <input
                            data-test="password"
                            onChange={[Function]}
                            placeholder="Password"
                            type="password"
                            value=""
              />
</FormField>,
            <Button
              as="button"
              color="teal"
              type="submit"
>
              Set new password
</Button>,
          ],
          "data-test": "form",
          "onSubmit": [Function],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": Array [
                <label>
                  New Password
</label>,
                <input
                  data-test="password"
                  onChange={[Function]}
                  placeholder="Password"
                  type="password"
                  value=""
/>,
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "New Password",
                },
                "ref": null,
                "rendered": "New Password",
                "type": "label",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "data-test": "password",
                  "onChange": [Function],
                  "placeholder": "Password",
                  "type": "password",
                  "value": "",
                },
                "ref": null,
                "rendered": null,
                "type": "input",
              },
            ],
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "as": "button",
              "children": "Set new password",
              "color": "teal",
              "type": "submit",
            },
            "ref": null,
            "rendered": "Set new password",
            "type": [Function],
          },
        ],
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <Header
            as="h1"
>
            Please set a new password
</Header>,
          null,
          <Form
            as="form"
            data-test="form"
            onSubmit={[Function]}
>
            <FormField>
                        <label>
                                    New Password
                        </label>
                        <input
                                    data-test="password"
                                    onChange={[Function]}
                                    placeholder="Password"
                                    type="password"
                                    value=""
                        />
            </FormField>
            <Button
                        as="button"
                        color="teal"
                        type="submit"
            >
                        Set new password
            </Button>
</Form>,
        ],
        "padded": "very",
        "size": "tiny",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "as": "h1",
            "children": "Please set a new password",
          },
          "ref": null,
          "rendered": "Please set a new password",
          "type": [Function],
        },
        null,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "as": "form",
            "children": Array [
              <FormField>
                <label>
                                New Password
                </label>
                <input
                                data-test="password"
                                onChange={[Function]}
                                placeholder="Password"
                                type="password"
                                value=""
                />
</FormField>,
              <Button
                as="button"
                color="teal"
                type="submit"
>
                Set new password
</Button>,
            ],
            "data-test": "form",
            "onSubmit": [Function],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": Array [
                  <label>
                    New Password
</label>,
                  <input
                    data-test="password"
                    onChange={[Function]}
                    placeholder="Password"
                    type="password"
                    value=""
/>,
                ],
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "New Password",
                  },
                  "ref": null,
                  "rendered": "New Password",
                  "type": "label",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "data-test": "password",
                    "onChange": [Function],
                    "placeholder": "Password",
                    "type": "password",
                    "value": "",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": "input",
                },
              ],
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "as": "button",
                "children": "Set new password",
                "color": "teal",
                "type": "submit",
              },
              "ref": null,
              "rendered": "Set new password",
              "type": [Function],
            },
          ],
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
